Add how to documentation for bignum test data generation (#52)
Mbed-TLS/mbedtls-docs: kb/how-to/add-test-generation.md

# How to systematically generate test data

> When adding new test functions or improving coverage of existing tests in Mbed TLS, a large number of test cases may need to be added. This can be time-consuming and error-prone when adding manually. Mbed TLS includes a Python framework which can be used to systematically generate test case data from input values. The framework is flexible, to support a variety of test functions, and uses inheritance to reduce code repetition in similar test functions.

Reviewer: I personally find long lines hard to read, even when the editor wraps them. Typically I find 90-100 characters a good maximum.

In addition, I also tend to start a new line when there's a new sentence - when this isn't done, a later PR changing the text will often cause reformatting of all the following lines, which makes diffs harder to read than they need to be.

Author: This is definitely an oversight, I'll add a commit to split the lines up.

## Adding new tests

There are two options for adding new tests:
 - Adding classes to an existing script, such as [tests/scripts/generate_bignum_tests.py](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/scripts/generate_bignum_tests.py).
 - Creating a new script.

When creating a new script, files will not automatically be generated when building tests unless the build system is updated.

## Creating a test script

This is only required if an existing script is not being extended. An example is included for each step, showing how to create a test generation script for `test_suite_mpi.function`, a basic equivalent of [generate_bignum_tests.py](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/scripts/generate_bignum_tests.py).

### Initial Python script

To create the script, `test_generation` must be imported from `mbedtls_dev` and `test_generation.main()` called when the script is ran. The `scripts/` directory must be added to the system path before importing from `mbedtls_dev`; if the new script is in `tests/scripts/` this can be done by importing `scripts_path`.

> For the example, create `generate_bignum_ex_tests.py` in `tests/scripts/`, containing:

Reviewer: Could we have a blank line here please?

```suggestion
For the example, create `generate_bignum_ex_tests.py` in `tests/scripts/`, containing:

```

 - Adding the script to
[/tests/scripts/check-generated-files.sh](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/scripts/check-generated-files.sh).
 - Adding the script to
[/scripts/make_generated_files.bat](https://github.com/Mbed-TLS/mbedtls/blob/development/scripts/make_generated_files.bat).
 - Adding calls and targets in CMake and Make for the script and its generated files.
The script will need to be called in the same places where `generate_bignum_tests.py` is
currently called.

## Generating test cases for a function

To add test cases for a function, a concrete class derived from `BaseTarget` must be
added to the script.
This class must implement all attributes and methods of `BaseTarget` and will generate
test cases.
This is a [test function class](#test-function-classes).

As an example, test case generation is added for the test function `mpi_add_mpi()`.
The added class will be derived from `BignumTarget`, and can be added either to
[tests/scripts/generate_bignum_tests.py](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/scripts/generate_bignum_tests.py)
or the example script described in the previous section.

> ### Creating the class

Reviewer 2: For the sake of people who aren't reading the document linearly, please clarify “the class” in the title. Even just “Creating the function class” would be a lot clearer.

The class must derive from a file target class, and the following class attributes should
be set:
 - `count = 0`: this resets the test case counting mechanism.
Alternatively, to disable test case numbering in descriptions, set
`show_test_count = False`.
 - `test_function`: the test function that the class will generate cases for.
 - `test_name`: a short descriptive name for the test, this can be the same as
`test_function` or a more readable equivalent.
This is used as the start of the description line.

For the example, add the following class:

```python
class BignumAddExample(BignumTarget):
    """Test cases for bignum addition."""
    count = 0
    test_function = "mpi_add_mpi"
    test_name = "MPI add"
```

### Class initialization

Each instance of the class will represent a different test case, each with unique inputs.
To initialize an instance of the class, these inputs must be passed to the `__init__()`
constructor method.

> For `mpi_add_mpi()`, two input values are required, `val_a` and `val_b`.

Reviewer 2: Almost everything in this paragraph is specific to generating bignum tests, and irrelevant otherwise. I think it would be more helpful to have a non-bignum or simplified example that doesn't go into these considerations about leading zeros, and a separate more advanced example that does illustrate how to deal with bignums. The more advanced example could be just a statement to go and look at a particular class in `generate_bignum_tests.py` though.

As the test function uses string inputs, the constructor will use the same format and
store the values as-is.
This ensures the values are correct in the generated data, preserving leading zeros and
null strings.
For calculating the expected result, these inputs are also converted to, and stored in,
`int` form; this is a multi-precision type in Python and so can be used for bignum
calculations.

For the example, add the following constructor to `BignumAddExample`:

```python
    def __init__(self, val_a: str, val_b: str) -> None:
        self.arg_a = val_a
        self.arg_b = val_b
        # int(val, 16) converts from a hex string to int
        # empty strings should be treated as zero in calculations
        self.int_a = int(val_a, 16) if val_a else 0
        self.int_b = int(val_b, 16) if val_b else 0
```

### Implementing abstract BaseTarget methods

> Abstract methods declared in `BaseTarget` must also be implemented in test function
> classes.

Reviewer 2: I'm not sure what that sentence means — I guess this?
```suggestion
The abstract methods declared in `BaseTarget` must have an implementation in test function
classes.
```
which could be simplified to
```suggestion
Test function classes must provide an implementation of all the abstract methods declared in `BaseTarget`.
```

For the example, the following imports will be required:

```python
import itertools
from typing import Iterator, List
# After import scripts_path
from mbedtls_dev import test_case
```

> `arguments()` returns the list of arguments required by the test function.

Reviewer 2: ```suggestion
The method `arguments()` returns the list of arguments passed to the test function.
```
(“required by” makes me think it returns some kind of specification that describes what the function requires, but that's not it.)

> Any C-string arguments must be quoted.

Reviewer 2: I'm not sure what “C-string arguments” refers to here. Is this about test function arguments that are passed to test functions as `char *`? If so this statement is incomplete: `data_t` arguments also need to be quoted. It would be clearer to formulate this differently, something like

> Arguments must have the syntax expected in `.data` files, with double quotes around arguments that are strings or hex data.

> In the example, a `result()` method is also added to calculate the expected output.

Reviewer 2: Suggested reformulation:
```suggestion
In the example, we also use an auxiliary method `result()` to calculate the expected output.
```
This makes it clearer that _we_ (the author(s)) are choosing to use a method called `result`, it isn't something that the framework requires.

This is a `classmethod`, which acts on the class, rather than a specific instance.
In the example, a list of strings will be used as inputs, and a test case is generated
for each possible combination.
The standard module `itertools` is used for generation of unique input combinations.

For the example, add the following method to `BignumAddExample`:

```python
    @classmethod
    def generate_function_tests(cls) -> Iterator[test_case.TestCase]:
        # Create a list of input values
        input_values = ["", "0", "1", "123"]
        # combinations_with_replacement generates a list of all unique
        # combinations, including where inputs are the same i.e. ("0", "0")
        for a_value, b_value in itertools.combinations_with_replacement(
            input_values, 2
        )
            # Initialize a class instance with these values
            class_instance = cls(a_value, b_value)
            # Yield the TestCase object from this instance
            yield class_instance.create_test_case()
```

### Other BaseTarget attributes and methods

`description()` returns a description of the test case, in the format "`test_name`

> #`count` `case_description`" by default.

Reviewer 2: Don't put `#` as the first character of a line: in many markdown variants, it starts a title even when the previous line is not blank.

`case_description` is an optional description for the specific test case, which may be
explicitly passed when constructing an instance, or generated in the class.
This can be used to add context, and in the example this will be generated by overriding
the `description()` method.

For the example, add the following method to `BignumAddExample`:

```python
    def description(self) -> str:
        if not self.case_description:
            self.case_description = "{} + {}".format(
                # hex() converts to a hex string with leading 0x
                "0 (null)" if self.arg_a == "" else hex(self.int_a),
                "0 (null)" if self.arg_b == "" else hex(self.int_b)
            )
        # Call description() in the parent class
        return super().description()
```

`dependencies` is a list of build macros, which specifies the required build config for
the test case to run.
This can be set as a class attribute if always required for the test function, or set

> conditionally in the constructor if dependent on the test case.

Reviewer 2: This warrants an example: how do you set a dependency for one specific test case?

(Setting a dependency for a whole function is rarely useful, since that can go in the `.function` file. It's only useful when there are multiple function classes with the same function name.)

The tests can then be generated and ran together with other suites by running `make test`.
To run only these tests:

```sh
$ python tests/scripts/generate_bignum_ex_tests.py
$ make tests/test_suite_mpi.ex.generated
$ cd tests
$ ./test_suite_mpi.ex.generated
```

## Extending existing test classes

When adding test case generation for functions, there may be previously implemented
classes which generate tests for similar functions.
For example, the test function `mpi_add_abs()` is very similar to `mpi_add_mpi()`, but
returns an absolute result.
Test structure, input format, and useful test cases may also be common across test
functions.
For example, a variety of test functions for binary operations use two inputs A and B,
and one output, and can be implemented similarly.

### Variants

For variants of a test function, inheritance can be used to reduce repetition of code.
For example, to implement the test function `mpi_add_abs()`, `BignumAddAbs` can inherit
from the `BignumAdd` class.
The necessary changes for implementing `BignumAddAbs` are then reduced:
 - `count` should be reset to 0 for the class.
 - `test_function` should be set to `mpi_add_mpi_abs`.

>  - `test_name` should be set.

Reviewer 2: ```suggestion
 - `test_name` should be set to a different name.
```

>  - `int` forms of the inputs (`int_a` and `int_b`) should be made absolute.

Reviewer 2: ```suggestion
 - We take the same list of inputs as the parent class, but calculate the expected result differently.
```

For example, possible implementation of this class:

```python
class BignumAddAbsExample(BignumAddExample):
    """Tests for absolute variant of bignum addition."""
    count = 0
    test_function = "mpi_add_abs"
    test_name = "MPI add (abs)"

# Either override __init__()
    def __init__(self, val_a, val_b) -> None:
        super().__init__(val_a, val_b)
        self.int_a = abs(self.int_a)

>         self.int_b = abs(self.int_b)

Reviewer 2: I guess this works due to the way `BignumAdd` is structured, because `__init__` stores the representation of the A and B arguments in fields that aren't `int_a` and `int_b` but does not itself calculate the representation of the expected-result argument. But if the reader doesn't know that, they'll be completely lost. Besides, it's confusing to show the same piece of code with three parts where you're only supposed to take parts 1+2 or 1+3. I suggest not showing this approach at all, or in a different example.
